src/Dictionaries/getDictionaryConfigurationFromAST.cpp

```diff
@@ -571,11 +571,24 @@ void checkPrimaryKey(const AttributeNameToConfiguration & all_attrs, const Names
 
 }
 
+void checkLifetime(const ASTCreateQuery & query)
+{
+    if (query.dictionary->layout && query.dictionary->layout->layout_type == "direct")
```

Comment:
It should be allowed if the dictionary is ATTACHed, e.g., during the server's startup. This is needed to allow seamless upgrades.

Reply:
Somewhere around `StorageFactory::Arguments`, `attach`, check its usage.